core/src/main/java/discord4j/core/object/command/Interaction.java

```diff
@@ -0,0 +1,178 @@
+    private final InteractionData data;
+
+    /**
+     * Constructs an {@code Interaction} with an associated {@link GatewayDiscordClient} and Discord data.
+     *
+     * @param gateway The {@link GatewayDiscordClient} associated to this object, must be non-null.
+     * @param data The raw data as represented by Discord, must be non-null.
+     */
+    public Interaction(final GatewayDiscordClient gateway, final InteractionData data) {
+        this.gateway = Objects.requireNonNull(gateway);
+        this.data = Objects.requireNonNull(data);
+    }
+
+    public InteractionData getData() {
+        return data;
+    }
+
+    /**
+     * Gets the id of the interaction.
+     *
+     * @return The id of the interaction.
+     */
+    public Snowflake getId() {
+        return Snowflake.of(data.id());
+    }
+
+    /**
+     * Gets the type of interaction.
+     *
+     * @return The type of interaction.
+     */
+    public Type getType() {
+        return Type.of(data.type());
+    }
+
+    /**
+     * Gets the command data payload.
+     *
+     * @return The command data payload.
+     */
+    public ApplicationCommandInteraction getCommandInteraction() {
+        return new ApplicationCommandInteraction(getClient(), data.data().get(),
+                getGuildId().map(Snowflake::asLong).orElse(null));
+    }
+
+    /**
+     * Gets the guild id it was sent from, if invoked in a guild.
+     *
+     * @return The guild id it was sent from, if invoked in a guild.
+     */
+    public Optional<Snowflake> getGuildId() {
+        return data.guildId().toOptional().map(Snowflake::of);
+    }
+
+    /**
+     * Gets the guild it was sent from, if invoked in a guild.
+     *
+     * @return The guild it was sent from, if invoked in a guild.
+     */
+    public Mono<Guild> getGuild() {
+        return Mono.justOrEmpty(getGuildId()).flatMap(gateway::getGuildById);
+    }
+
+    /**
+     * Gets the channel id it was sent from.
+     *
+     * @return The channel id it was sent from.
+     */
+    public Snowflake getChannelId() {
+        return Snowflake.of(data.channelId().get());
+    }
+
+    /**
+     * Gets the channel it was sent from.
+     *
+     * @return The channel it was sent from.
+     */
+    public Mono<TextChannel> getChannel() {
+        return gateway.getChannelById(getChannelId()).cast(TextChannel.class);
+    }
+
+    /**
+     * Gets the invoking member, if invoked in a guild.
+     *
+     * @return The invoking member, if invoked in a guild.
+     */
+    public Optional<Member> getMember() {
+        return data.member().toOptional()
+                .map(data -> new Member(gateway, data, getGuildId().get().asLong()));
+    }
+
+    /**
+     * Gets the invoking user.
+     *
+     * @return The invoking user.
+     */
+    public User getUser() {
+        UserData userData = data.member().isAbsent() ? data.user().get() : data.member().get().user();
+        return new User(getClient(), userData);
+    }
```

Comment:
This is fine. Just add a note in docs that we'll get the data from member for guild interactions.